aws/resource_aws_codestarnotifications_notification_rule.go

```diff
@@ -0,0 +1,291 @@
+		Targets:      expandCodeStarNotificationsNotificationRuleTargets(d.Get("target").(*schema.Set).List()),
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		params.Tags = keyvaluetags.New(v.(map[string]interface{})).IgnoreAws().CodestarnotificationsTags()
+	}
+
+	res, err := conn.CreateNotificationRule(params)
+	if err != nil {
+		return fmt.Errorf("error creating codestar notification rule: %s", err)
+	}
+
+	d.SetId(aws.StringValue(res.Arn))
+
+	return resourceAwsCodeStarNotificationsNotificationRuleRead(d, meta)
+}
+
+func resourceAwsCodeStarNotificationsNotificationRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codestarnotificationsconn
+
+	rule, err := conn.DescribeNotificationRule(&codestarnotifications.DescribeNotificationRuleInput{
+		Arn: aws.String(d.Id()),
+	})
+
+	if err != nil {
+		if isAWSErr(err, codestarnotifications.ErrCodeResourceNotFoundException, "") {
+			log.Printf("[WARN] codestar notification rule (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error reading codestar notification rule: %s", err)
+	}
+
+	d.Set("arn", rule.Arn)
+	d.Set("detail_type", rule.DetailType)
+	eventTypeIds := make([]string, 0, len(rule.EventTypes))
+	for _, et := range rule.EventTypes {
+		eventTypeIds = append(eventTypeIds, aws.StringValue(et.EventTypeId))
+	}
+	if err := d.Set("event_type_ids", eventTypeIds); err != nil {
+		return fmt.Errorf("error setting event_type_ids: %s", err)
+	}
+	d.Set("name", rule.Name)
+	d.Set("status", rule.Status)
+	d.Set("resource", rule.Resource)
+	d.Set("tags", keyvaluetags.New(rule.Tags).IgnoreAws().Map())
+
+	targets := make([]map[string]interface{}, 0, len(rule.Targets))
+	for _, t := range rule.Targets {
+		targets = append(targets, map[string]interface{}{
+			"address": aws.StringValue(t.TargetAddress),
+			"type":    aws.StringValue(t.TargetType),
+			"status":  aws.StringValue(t.TargetStatus),
+		})
+	}
+	if err = d.Set("target", targets); err != nil {
+		return fmt.Errorf("error setting codestar notification target: %s", err)
+	}
+
+	return nil
+}
+
+const awsCodeStartNotificationsNotificationRuleErrorSubscribed = "The target cannot be deleted because it is subscribed to one or more notification rules."
+
+// cleanupCodeStarNotificationsNotificationRuleTargets tries to remove unused notification targets. AWS API does not
+// provide expicit way for creating targets, they are created on first subscription. Here we are trying to remove all
+// unused targets which were unsubscribed from this notification rule.
+func cleanupCodeStarNotificationsNotificationRuleTargets(conn *codestarnotifications.CodeStarNotifications, oldVal *schema.Set, newVal *schema.Set) error {
+	removedTargets := oldVal
+	if newVal != nil {
+		removedTargets = oldVal.Difference(newVal)
+	}
+
+	targets := removedTargets.List()
+
+	err := resource.Retry(time.Duration(5)*time.Second, func() *resource.RetryError {
```

Comment:
This logic may be problematic in practice (https://github.com/terraform-providers/terraform-provider-aws/issues/7873 / https://github.com/hashicorp/terraform-plugin-sdk/pull/199), however it is more valuable to get this new resource out so folks can start trying it.